File: server/src/routes/match.js
import { Router } from 'express';
import multer from 'multer';
import { v4 as uuid } from 'uuid';
import db from '../db.js';
import { extractText } from '../services/fileParser.js';
import { matchResumesHybrid } from '../services/hybridMatcher.js';
import {
  embedBatch,
  vectorToBuffer,
  bufferToVector,
  buildResumeEmbeddingText,
} from '../services/embeddingService.js';

// Check if OpenAI API key is configured
function isAIEnabled() {
  return !!process.env.OPENAI_API_KEY;
}

const router = Router();

const MAX_MB = parseInt(process.env.MAX_FILE_SIZE_MB || '10', 10);

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: MAX_MB * 1024 * 1024 },
  fileFilter: (req, file, cb) => {
    const ext = '.' + file.originalname.split('.').pop().toLowerCase();
    ['.pdf', '.docx', '.doc', '.txt'].includes(ext) ? cb(null, true) : cb(new Error(`File type not allowed: ${ext}`));
  },
});

/**
 * POST /api/match
 *
 * Multipart form fields:
 *   - requirements (file) OR requirementsText (string)
 *   - resumes[]  (files, up to 50)
 *   - jobId      (optional — if provided, uses stored job embedding)
 *   - jobTitle   (optional)
 *   - preferences (JSON string)
 */
router.post(
  '/',
  upload.fields([
    { name: 'requirements', maxCount: 1 },
    { name: 'resumes',      maxCount: 50 },
  ]),
  async (req, res) => {
    try {
      // ── 1. Requirements text ───────────────────────────────────────────────
      let requirementsText = req.body.requirementsText || '';

      if (!requirementsText && req.files?.requirements?.[0]) {
        const f = req.files.requirements[0];
        requirementsText = await extractText(f.buffer, f.originalname);
      }

      // If a jobId is provided, load the stored description as fallback
      const jobId = req.body.jobId || null;
      let jobRow  = null;
      if (jobId) {
        jobRow = db.prepare('SELECT * FROM jobs WHERE id = ?').get(jobId);
        if (jobRow && !requirementsText) {
          requirementsText = jobRow.description;
        }
      }

      if (!requirementsText?.trim()) {
        return res.status(400).json({ error: 'Job requirements are required (text or file).' });
      }

      // ── 2. Parse resumes ───────────────────────────────────────────────────
      // TODO: Add validation to check if uploaded file is actually a resume vs job description
      // TODO: Implement document type detection (resume indicators: contact info, experience, education vs JD indicators: requirements, responsibilities)
      // TODO: Reject non-resume files with clear error messages
      const resumeFiles = req.files?.resumes || [];
      if (resumeFiles.length === 0) {
        return res.status(400).json({ error: 'At least one resume file is required.' });
      }

      const parseErrors = [];
      const parsedResumes = (
        await Promise.all(
          resumeFiles.map(async (file) => {
            try {
              const text = await extractText(file.buffer, file.originalname);
              // TODO: Validate that extracted text looks like a resume (check for resume-specific sections)
              const name = file.originalname.replace(/\.[^.]+$/, '').replace(/[-_]/g, ' ');
              return { name, fileName: file.originalname, fileSize: file.size, text };
            } catch (err) {
              parseErrors.push({ file: file.originalname, error: err.message });
              return null;
            }
          })
        )
      ).filter(Boolean);

      if (parsedResumes.length === 0) {
        return res.status(400).json({ error: 'Could not parse any resume files.', details: parseErrors });
      }

      // ── 3. Preferences ─────────────────────────────────────────────────────
      let preferences = {};
      try { preferences = req.body.preferences ? JSON.parse(req.body.preferences) : {}; }
      catch { return res.status(400).json({ error: 'Invalid preferences JSON.' }); }

      const prefs = {
        skillsWeight:       Number(preferences.skillsWeight      ?? 40),
        experienceWeight:   Number(preferences.experienceWeight   ?? 25),
        educationWeight:    Number(preferences.educationWeight    ?? 20),
        overallWeight:      Number(preferences.overallWeight      ?? 15),
      };

      // ── 4. Get job embedding for semantic matching ────────────────────────
      let jobVector = null;
      if (jobRow?.embedding) {
        jobVector = bufferToVector(jobRow.embedding);
        console.log(`Using stored job embedding (${jobVector.length} dims) for "${jobRow.title}"`);
      }

      // ── 5. Run Hybrid matching ────────────────────────────────────────────
      console.log('Using Hybrid matcher (Rule-based + Embeddings)');
      
      let results;
      let scoringMethod = 'hybrid';

      try {
        results = await matchResumesHybrid(
          requirementsText,
          parsedResumes,
          prefs,
          jobVector
        );
      } catch (err) {
        console.error('Hybrid matching error:', err);
        return res.status(500).json({ error: 'Matching failed: ' + err.message });
      }

      // ── 6. Store resume embeddings for future use ─────────────────────────
      let resumeEmbeddings = null;
      if (isAIEnabled()) {
        try {
          const texts = parsedResumes.map(buildResumeEmbeddingText);
          resumeEmbeddings = await embedBatch(texts);
          console.log(`✓ Resume embeddings generated for storage`);
        } catch (err) {
          console.error('Resume embedding failed:', err.message);
        }
      }

      // ── 7. Persist session + candidates ───────────────────────────────────
      const sessionId = uuid();
      const jobTitle  = req.body.jobTitle || jobRow?.title || 'Untitled Session';

      // Only use jobId if the job actually exists in the database
      const validJobId = jobRow ? jobId : null;

      db.prepare(`
        INSERT INTO sessions (id, job_id, job_title, preferences, result_count)
        VALUES (?, ?, ?, ?, ?)
      `).run(sessionId, validJobId, jobTitle, JSON.stringify(prefs), results.length);

      const insertCandidate = db.prepare(`
        INSERT INTO candidates
          (id, session_id, name, file_name, file_size, raw_text,
           final_score, grade_label, grade_color, breakdown, 
           title, location, email, phone, strengths, weaknesses, summary,
           embedding, embedding_model)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      `);

      db.transaction((candidates) => {
        candidates.forEach((c, i) => {
          const resumeIdx = parsedResumes.findIndex((r) => r.fileName === c.fileName);
          const resume    = parsedResumes[resumeIdx];
          const vec       = resumeEmbeddings?.[resumeIdx];

          insertCandidate.run(
            uuid(), sessionId, c.name, c.fileName,
            resume?.fileSize || 0,
            resume?.text?.slice(0, 5000) || '',
            c.finalScore, c.grade.label, c.grade.color,
            JSON.stringify(c.breakdown),
            c.title || null,
            c.location || null,
            c.email || null,
            c.phone || null,
            JSON.stringify(c.strengths || []),
            JSON.stringify(c.weaknesses || []),
            c.summary || null,
            vec ? vectorToBuffer(vec) : null,
            vec ? 'text-embedding-3-small' : null,
          );
        });
      })(results);

      // ── 8. Respond ─────────────────────────────────────────────────────────
      res.json({
        sessionId,
        jobTitle,
        scoredBy:        scoringMethod,
        semanticEnabled: !!jobVector,
        totalParsed:     parsedResumes.length,
        parseErrors,
        results,
      });

    } catch (err) {
      console.error('Match error:', err);
      res.status(500).json({ error: err.message || 'Internal server error' });
    }
  }
);

function getGrade(score) {
  if (score >= 85) return { label: 'Excellent',     color: 'emerald' };
  if (score >= 70) return { label: 'Good',          color: 'blue' };
  if (score >= 55) return { label: 'Fair',          color: 'yellow' };
  if (score >= 40) return { label: 'Below Average', color: 'orange' };
  return              { label: 'Poor',          color: 'red' };
}

export default router;

File: server_py/routes/match.py
"""
routes/match.py — Resume matching endpoint.

Endpoint:
  POST /api/match  — accepts multipart form with:
    - resumes[]          one or more resume files (PDF/DOCX/TXT, up to 50)
    - requirements       optional JD file
    - requirementsText   optional JD plain text
    - jobId               optional saved job ID (loads stored embedding)
    - jobTitle            optional session label
    - preferences         JSON weights for skills / experience / education / overall

Flow: extract text → hybrid match (NLP + semantic) → persist session &
candidates to SQLite → return ranked results with session ID.
"""

import json
import uuid
from fastapi import APIRouter, UploadFile, File, Form, HTTPException
from db import db
from services.file_parser import extract_text
from services.hybrid_matcher import match_resumes_hybrid
from services.embedding_service import embed_batch, vector_to_bytes, bytes_to_vector, build_resume_embedding_text
import os

router = APIRouter()

MAX_CUSTOM_CRITERIA = 5
MAX_TERM_LENGTH = 150


def sanitize_custom_criteria(raw):
    if not isinstance(raw, list):
        return []
    out = []
    for i, c in enumerate(raw):
        if not isinstance(c, dict):
            continue
        term = str(c.get("term", "")).strip()
        if not term or c.get("enabled") is False:
            continue
        try:
            weight = float(c.get("weight", 0))
        except (TypeError, ValueError):
            weight = 0
        weight = max(0, min(50, weight))
        if weight <= 0:
            continue
        out.append({"id": c.get("id") or f"custom_{i}", "term": term[:MAX_TERM_LENGTH], "weight": weight})
        if len(out) >= MAX_CUSTOM_CRITERIA:
            break
    return out

@router.post("")
async def match_resumes(
    resumes: list[UploadFile] = File(...),
    requirements: UploadFile = File(None),
    requirementsText: str = Form(""),
    jobId: str = Form(""),
    jobTitle: str = Form(""),
    preferences: str = Form("{}"),
):
    conn = db()

    # 1. Requirements text
    req_text = requirementsText
    if not req_text and requirements:
        data = await requirements.read()
        req_text = extract_text(data, requirements.filename)

    job_row = None
    if jobId:
        job_row = conn.execute("SELECT * FROM jobs WHERE id=?", (jobId,)).fetchone()
        if job_row and not req_text:
            req_text = dict(job_row)["description"]

    if not req_text or not req_text.strip():
        raise HTTPException(400, "Job requirements are required (text or file).")

    # 2. Parse resumes
    if not resumes:
        raise HTTPException(400, "At least one resume file is required.")

    parse_errors = []
    parsed_resumes = []
    for f in resumes:
        try:
            data = await f.read()
            text = extract_text(data, f.filename)
            name = f.filename.rsplit(".", 1)[0].replace("-", " ").replace("_", " ")
            parsed_resumes.append({"name": name, "fileName": f.filename, "fileSize": len(data), "text": text})
        except Exception as e:
            parse_errors.append({"file": f.filename, "error": str(e)})

    if not parsed_resumes:
        raise HTTPException(400, "Could not parse any resume files.")

    # 3. Preferences
    try:
        prefs_raw = json.loads(preferences)
    except Exception:
        raise HTTPException(400, "Invalid preferences JSON.")
    prefs = {
        "skillsWeight":     float(prefs_raw.get("skillsWeight", 40)),
        "experienceWeight": float(prefs_raw.get("experienceWeight", 25)),
        "educationWeight":  float(prefs_raw.get("educationWeight", 20)),
        "overallWeight":    float(prefs_raw.get("overallWeight", 15)),
        "customCriteria":   sanitize_custom_criteria(prefs_raw.get("customCriteria")),
    }

    # 4. Job embedding
    job_vector = None
    if job_row and dict(job_row).get("embedding"):
        job_vector = bytes_to_vector(dict(job_row)["embedding"])

    # 5. Match
    results = await match_resumes_hybrid(req_text, parsed_resumes, prefs, job_vector)

    # 6. Resume embeddings for storage
    resume_embeddings = None
    if os.getenv("OPENAI_API_KEY"):
        try:
            texts = [build_resume_embedding_text(r) for r in parsed_resumes]
            resume_embeddings = await embed_batch(texts)
        except Exception as e:
            print(f"Resume embedding failed: {e}")

    # 7. Persist session + candidates
    session_id = str(uuid.uuid4())
    final_job_title = jobTitle or (dict(job_row)["title"] if job_row else "Untitled Session")
    valid_job_id = jobId if job_row else None

    conn.execute(
        "INSERT INTO sessions (id,job_id,job_title,preferences,result_count) VALUES (?,?,?,?,?)",
        (session_id, valid_job_id, final_job_title, json.dumps(prefs), len(results)),
    )

    for i, c in enumerate(results):
        res_idx = next((j for j, r in enumerate(parsed_resumes) if r["fileName"] == c["fileName"]), None)
        resume = parsed_resumes[res_idx] if res_idx is not None else {}
        vec = resume_embeddings[res_idx] if resume_embeddings and res_idx is not None else None
        conn.execute(
            """INSERT INTO candidates
                (id,session_id,name,file_name,file_size,raw_text,
                final_score,grade_label,grade_color,
                breakdown,title,location,email,phone,
                strengths,weaknesses,summary,
                interview_focus_areas,interview_questions,top_interview_questions,
                embedding,embedding_model)
                VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)""",
            (
                str(uuid.uuid4()), session_id, c["name"], c["fileName"],
                resume.get("fileSize", 0),
                (resume.get("text") or ""),
                c["finalScore"], c["grade"]["label"], c["grade"]["color"],
                json.dumps(c["breakdown"]),
                c.get("title"), c.get("location"), c.get("email"), c.get("phone"),
                json.dumps(c.get("strengths", [])),
                json.dumps(c.get("weaknesses", [])),
                c.get("summary"),
                json.dumps(c.get("interviewFocusAreas", [])),
                json.dumps(c.get("interviewQuestions", [])),
                json.dumps(c.get("topInterviewQuestions", [])),
                vector_to_bytes(vec) if vec else None,
                "text-embedding-3-small" if vec else None,
            ),
        )
    conn.commit()

    return {
        "sessionId": session_id,
        "jobTitle": final_job_title,
        "scoredBy": results[0].get("scoredBy", "hybrid") if results else "hybrid",
        "semanticEnabled": bool(job_vector),
        "totalParsed": len(parsed_resumes),
        "parseErrors": parse_errors,
        "results": results,
    }